src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs

```diff
@@ -0,0 +1,244 @@
+        /// 
+        /// Unless <see cref="MetadataStreamOptions.PrefetchMetadata"/> is specified no data 
+        /// is read from the stream during the construction of the <see cref="MetadataReaderProvider"/>. Furthermore, the stream must not be manipulated
+        /// by caller while the <see cref="MetadataReaderProvider"/> is alive and undisposed.
+        /// 
+        /// If <see cref="MetadataStreamOptions.PrefetchMetadata"/>, the <see cref="MetadataReaderProvider"/> 
+        /// will have read all of the data requested during construction. As such, if <see cref="MetadataStreamOptions.LeaveOpen"/> is also
+        /// specified, the caller retains full ownership of the stream and is assured that it will not be manipulated by the <see cref="MetadataReaderProvider"/>
+        /// after construction.
+        /// </param>
+        /// <exception cref="ArgumentOutOfRangeException">Size is negative or extends past the end of the stream.</exception>
+        public static MetadataReaderProvider FromMetadataStream(Stream stream, MetadataStreamOptions options = MetadataStreamOptions.Default, int size = 0)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            if (!stream.CanRead || !stream.CanSeek)
+            {
+                throw new ArgumentException(SR.StreamMustSupportReadAndSeek, nameof(stream));
+            }
+
+            if (!options.IsValid())
+            {
+                throw new ArgumentOutOfRangeException(nameof(options));
+            }
+
+            long start = stream.Position;
+            int actualSize = StreamExtensions.GetAndValidateSize(stream, size, nameof(stream));
+
+            MetadataReaderProvider result;
+            bool closeStream = true;
+            try
+            {
+                bool isFileStream = FileStreamReadLightUp.IsFileStream(stream);
+
+                if ((options & MetadataStreamOptions.PrefetchMetadata) == 0)
+                {
+                    result = new MetadataReaderProvider(new StreamMemoryBlockProvider(stream, start, actualSize, isFileStream, (options & MetadataStreamOptions.LeaveOpen) != 0));
+                    closeStream = false;
+                }
+                else
+                {
+                    // Read in the entire image or metadata blob:
+                    result = new MetadataReaderProvider(StreamMemoryBlockProvider.ReadMemoryBlockNoLock(stream, isFileStream, start, actualSize));
+                   
+                    // We read all we need, the stream is going to be closed.
+                }
+            }
+            finally
+            {
+                if (closeStream && (options & MetadataStreamOptions.LeaveOpen) == 0)
+                {
+                    stream.Dispose();
+                }
+            }
+
+            return result;
+        }
+
```

Comment:
Is FromPeStream as discussed offline planned for later? Also wonder about a version that can sniff any format with embedded metadata (not very common use case, but perhaps nice to prevent code duplication between visualizer tools, etc.?)